src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.InteropServices;
+#if IS_DESKTOP
+using System.Security.Cryptography;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using NuGet.Common;
+#endif
+
+namespace NuGet.Packaging.Signing
+{
+    internal static class NativeUtilities
+    {
+        internal static void SafeFree(IntPtr ptr)
+        {
+            if (ptr != IntPtr.Zero)
+            {
+                Marshal.FreeHGlobal(ptr);
+            }
+        }
+
+        internal static void ThrowIfFailed(bool result)
+        {
+            if (!result)
+            {
+                Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+            }
+        }
+#if IS_DESKTOP
+        internal static SignedCms NativeSign(byte[] data, X509Certificate2 certificate, CngKey privateKey, CryptographicAttributeObjectCollection attributes, Common.HashAlgorithmName hashAlgorithm)
+        {
+            using (var hb = new HeapBlockRetainer())
+            using (var chain = new X509Chain())
+            {
+                // Skipping a bunch of chain building options we should normally set otherwise
+                chain.Build(certificate);
```

Comment:
Some options need to be set, some additional certs need to be added, and the return result must be checked.